zkir/Dialect/Field/Conversions/FieldToModArith/Extension/QuarticExtensionField.cpp

```diff
@@ -0,0 +1,471 @@
+/* Copyright 2025 The ZKIR Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "zkir/Dialect/Field/Conversions/FieldToModArith/Extension/QuarticExtensionField.h"
+
+#include "zkir/Dialect/Field/Conversions/FieldToModArith/ConversionUtils.h"
+#include "zkir/Dialect/Field/IR/FieldAttributes.h"
+#include "zkir/Dialect/ModArith/IR/ModArithOps.h"
+#include "zkir/Utils/APIntUtils.h"
+
+namespace mlir::zkir::field {
+
+QuarticExtensionField::QuarticExtensionField(ImplicitLocOpBuilder &b,
+                                             ExtensionFieldTypeInterface type,
+                                             const TypeConverter *converter)
+    : ExtensionField(b, type, converter) {
+  // Get the modulus from the base field type.
+  auto baseField = cast<PrimeFieldType>(type.getBaseFieldType());
+  llvm::APInt modulus = baseField.getModulus().getValue();
+  auto convertedType = converter->convertType(baseField);
+  unsigned bitWidth = modulus.getBitWidth();
+
+  // Precompute inverse constants for Toom-Cook algorithm.
+  auto createInvConst = [&](uint64_t n) {
+    llvm::APInt nVal(bitWidth, n);
+    nVal = nVal.urem(modulus);
+    auto constOp = b.create<mod_arith::ConstantOp>(
+        convertedType, IntegerAttr::get(baseField.getStorageType(), nVal));
+    return b.create<mod_arith::InverseOp>(constOp);
+  };
+
+  auto createConst = [&](uint64_t n) {
+    llvm::APInt nVal(bitWidth, n);
+    return b.create<mod_arith::ConstantOp>(
+        convertedType, IntegerAttr::get(baseField.getStorageType(), nVal));
+  };
+
+  zero = createConst(0);
+  inv2 = createInvConst(2);
+  inv3 = createInvConst(3);
+  inv4 = createInvConst(4);
+  inv6 = createInvConst(6);
+  inv12 = createInvConst(12);
+  inv20 = createInvConst(20);
+  inv24 = createInvConst(24);
+  inv30 = createInvConst(30);
+  inv120 = createInvConst(120);
+  const3 = createConst(3);
+  const5 = createConst(5);
+  const7 = createConst(7);
+  const9 = createConst(9);
+  const12 = createConst(12);
+  const15 = createConst(15);
+  const27 = createConst(27);
+
+  // Precompute Frobenius coefficients.
+  // For v⁴ = ξ, we have vᵏ maps to vᵏ⁽ᵖᵉ ᵐᵒᵈ ⁴⁾ * ξᵏ⁽ᵖᵉ ᵈⁱᵛ ⁴⁾.
```

Comment:
Isn't it this? Where does p and e come from?

```suggestion
  // For v⁴ = ξ, we have vᵏ maps to ξ^(k / 4) * v^(k % 4)
```